File: examples/footguns/12_deadlock/main.go
// Footgun: Deadlock — Go
//
// Go's sync.Mutex blocks the goroutine forever on deadlock. The runtime
// detects when ALL goroutines are blocked (a "global" deadlock) and panics
// with "all goroutines are asleep — deadlock!". But it does NOT detect
// partial deadlocks (a subset of goroutines deadlocked while others run).
//
// In practice, server programs always have goroutines running (the HTTP
// listener, timers, etc.), so a partial deadlock hangs silently — the server
// stops serving some requests but keeps running. The only evidence is in
// `go tool pprof` goroutine dumps or runtime/pprof block profiling.
//
// Go's race detector (-race) does not detect deadlocks, only data races.

package main

import (
	"fmt"
	"sync"
	"time"
)

// BROKEN: AB / BA lock order — deadlock.
// Not run in main(); shown for illustration.
func broken() {
	var muA, muB sync.Mutex

	go func() {
		muA.Lock()
		fmt.Println("g1: acquired A")
		time.Sleep(1 * time.Millisecond)
		fmt.Println("g1: waiting for B...")
		muB.Lock() // blocks: g2 holds B
		fmt.Println("g1: acquired B (never reached)")
		muB.Unlock()
		muA.Unlock()
	}()

	go func() {
		muB.Lock()
		fmt.Println("g2: acquired B")
		time.Sleep(1 * time.Millisecond)
		fmt.Println("g2: waiting for A...")
		muA.Lock() // blocks: g1 holds A — DEADLOCK
		fmt.Println("g2: acquired A (never reached)")
		muA.Unlock()
		muB.Unlock()
	}()

	// These joins would hang:
	// wg.Wait()
}

// CORRECT: consistent lock order — always A before B.
func correct() {
	var muA, muB sync.Mutex
	var wg sync.WaitGroup

	transfer := func(name string) {
		defer wg.Done()
		muA.Lock()   // always A first
		muB.Lock()   // then B
		fmt.Printf("%s: holds both locks\n", name)
		muB.Unlock()
		muA.Unlock()
	}

	wg.Add(2)
	go transfer("g1")
	go transfer("g2")
	wg.Wait()
}

// CORRECT: avoid multi-lock patterns via channels.
// Channels carry both the value and the synchronization — no explicit locks.
func withChannels() {
	type transfer struct{ from, to int; amount int }

	ch := make(chan transfer, 1)
	balances := map[int]int{1: 100, 2: 200}

	// Single goroutine owns the balances map — no lock needed.
	go func() {
		for t := range ch {
			balances[t.to] += t.amount
			balances[t.from] -= t.amount
		}
	}()

	ch <- transfer{from: 1, to: 2, amount: 50}
	ch <- transfer{from: 2, to: 1, amount: 30}
	close(ch)

	// Give the goroutine time to process (in real code: sync via done channel).
	time.Sleep(time.Millisecond)
	fmt.Printf("balances: %v\n", balances)
}

func main() {
	fmt.Println("--- consistent lock order (safe) ---")
	correct()

	fmt.Println("--- channel-based (no locks needed) ---")
	withChannels()

	fmt.Println("--- broken (not run — would deadlock) ---")
	// broken() // do not call
}

// Key insight: Go detects GLOBAL deadlock (all goroutines blocked) but not
// PARTIAL deadlock (subset blocked). Production services almost always have
// other goroutines running, so partial deadlocks silently hang.
// The idiomatic Go fix: avoid multi-lock patterns by using channels and
// having a single goroutine own each shared resource.

File: examples/footguns/12_deadlock/main.c
/*
 * Footgun: Deadlock — C
 *
 * C's pthread mutexes block forever when a deadlock occurs. There is no
 * built-in detection or timeout. The process hangs silently; the only
 * indication is that it stops making progress.
 *
 * The classic pattern: two threads, two mutexes, opposite acquisition order.
 * Thread A holds mutex1, waits for mutex2.
 * Thread B holds mutex2, waits for mutex1.
 * Neither can proceed. The program hangs indefinitely.
 *
 * Detection requires external tooling (Helgrind, ThreadSanitizer with
 * -fsanitize=thread, or gdb with thread apply all bt). There is no language
 * construct that prevents this.
 */

#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

static pthread_mutex_t mu_a = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t mu_b = PTHREAD_MUTEX_INITIALIZER;

/* Thread 1: locks A then B. */
static void *thread1(void *arg) {
    (void)arg;
    pthread_mutex_lock(&mu_a);
    printf("thread1: acquired A\n");
    usleep(1000); /* give thread2 time to acquire B */
    printf("thread1: waiting for B...\n");
    pthread_mutex_lock(&mu_b);  /* DEADLOCK: thread2 holds B */
    printf("thread1: acquired B (never reached)\n");
    pthread_mutex_unlock(&mu_b);
    pthread_mutex_unlock(&mu_a);
    return NULL;
}

/* Thread 2: locks B then A — opposite order. */
static void *thread2(void *arg) {
    (void)arg;
    pthread_mutex_lock(&mu_b);
    printf("thread2: acquired B\n");
    usleep(1000); /* give thread1 time to acquire A */
    printf("thread2: waiting for A...\n");
    pthread_mutex_lock(&mu_a);  /* DEADLOCK: thread1 holds A */
    printf("thread2: acquired A (never reached)\n");
    pthread_mutex_unlock(&mu_a);
    pthread_mutex_unlock(&mu_b);
    return NULL;
}

/* CORRECT: consistent lock order — always A before B. */
static void *thread_fixed(void *arg) {
    (void)arg;
    /* Both threads take A first, then B — no cycle possible. */
    pthread_mutex_lock(&mu_a);
    pthread_mutex_lock(&mu_b);
    printf("fixed thread: holds both locks\n");
    pthread_mutex_unlock(&mu_b);
    pthread_mutex_unlock(&mu_a);
    return NULL;
}

int main(void) {
    printf("--- correct: consistent lock order ---\n");
    pthread_t t3, t4;
    pthread_create(&t3, NULL, thread_fixed, NULL);
    pthread_create(&t4, NULL, thread_fixed, NULL);
    pthread_join(t3, NULL);
    pthread_join(t4, NULL);

    printf("--- broken: opposite lock order (will deadlock) ---\n");
    printf("(not run — would hang the program)\n");
    /* pthread_t t1, t2;
     * pthread_create(&t1, NULL, thread1, NULL);
     * pthread_create(&t2, NULL, thread2, NULL);
     * pthread_join(t1, NULL);  -- hangs here
     * pthread_join(t2, NULL); */

    return 0;
}

/*
 * Compile: gcc -o deadlock_c main.c -lpthread && ./deadlock_c
 *
 * Detect the broken version: gcc -fsanitize=thread -o deadlock_c main.c -lpthread
 * (TSan detects the lock cycle and reports it before the hang.)
 *
 * Fix: enforce a global lock acquisition order. If locks have a natural
 * identity (e.g., pointer address), always acquire the lower-address lock
 * first: if (&mu_a < &mu_b) { lock a then b } else { lock b then a }
 */
